### ISSUE TITLE
The generics don't work with the type of a key value object

### ISSUE THREAD

**Issue by zabirauf** [Microsoft/TypeScript#16957]
**TypeScript Version:**  2.4.0 / nightly (2.5.0-dev.201xxxxx)

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
function test<T extends string | number>(a: T): {[key: T]: boolean} {
    const map: { [key: T]: boolean } = {};
    return map;
}
```

**Expected behavior:**
The type system should not complain that the key type of `map` should be string or number.

**Actual behavior:**
It gives error that the key of `map` should be `string or number` even though T is string or number

**Comment by snewell92** [comment]
Possibly related

```typescript
type Key = string | number;

const dict: { [index: Key]: any } = {};
```

**Comment by mhegazy** [comment]
duplicate of https://github.com/Microsoft/TypeScript/issues/16760

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
